File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float('inf')
        max_profit = 0

        for price in prices:
            if price < min_price:
                min_price = price  # Update the lowest price so far
            elif price - min_price > max_profit:
                max_profit = price - min_price  # Update max profit if current profit is higher

        return max_profit

'''
float('inf') means infinity.
It's a built-in way to represent a number that's larger than any other number.
There is also float('-inf') for negative infinity.

Note:
In the line `if price < min_price:`
This means that if the current price is smaller than the lowest price we've seen so far, we update it.
If the price is higher, we check if selling now would give us a better profit.
'''

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.c
int maxProfit(int* prices, int pricesSize) {
    int minPrice = prices[0];
    int maxProfit = 0;

    for (int i = 1; i < pricesSize; i++) {
        // 더 싼 가격을 발견하면 매수 가격 갱신
        if (prices[i] < minPrice) {
            minPrice = prices[i];
        }
        // 현재 가격에 팔았을 때의 이익으로 최대 이익 갱신
        else {
            int profit = prices[i] - minPrice;

            if (profit > maxProfit) {
                maxProfit = profit;
            }
        }
    }

    return maxProfit;
}